process: rewrite process management

rogpeppe opened the pull request:
We rewrite the process management to use a more
disciplined approach to concurrency. There's a main
goroutine for each process that monitors any process
that's currently active, and another goroutine that's
responsible for coordinating updates to the configuration.

Logging isn't hooked up and restarts aren't tested at all,
but this is probably a good stopping point for a review.

rogpeppe committed:
process: rewrite process management

We rewrite the process management to use a more
disciplined approach to concurrency. There's a main
goroutine for each process that monitors any process
that's currently active, and another goroutine that's
responsible for coordinating updates to the configuration.

Logging isn't hooked up and restarts aren't tested at all,
but this is probably a good stopping point for a review.

rogpeppe committed:
wire up logging

We refactor the old logging code so that the in-memory ring buffer is
integrated with the rest of the logging, which enables us to fix the issue
with the old code that log messages could be lost between producing the
backlog and following the rest of the log.

rogpeppe committed:
process: make process info work

Also remove some vestiges of groups from the client and the
unused process description field.

rogpeppe committed:
config: better topological sorting

We change the algorithm so that we can start and stop
dependencies in parallel. This means that we'll be able
to use topological sorting for `StopAll` without having
to stop and wait for each dependency sequentially.

rogpeppe committed:
process: update process state before starting

The watcher wasn't finding the process in the notifier
because the notifier hadn't been told about it.

Also add some logging to help us debug issues like this
and factor out Manager.stopProcesses in an abortive
effort to make `StopAll` and friends stop in topological order
(it's not as straightforward as it seemed at first)

rogpeppe committed:
cmd/gopm: tests for logging to a file

rogpeppe committed:
process: implement and test restarting

rogpeppe committed:
supervisor: shut down gracefully

This is not only the right thing to do, it also allows us to see coverage
information (about 70% of the process package, so not too bad)
because we're no longer using `os.Exit` to exit without the
process being able to write its coverage info.

rogpeppe committed:
cmd/gopmctl: working status command

rogpeppe commented on supervisor.go:
```diff
@@ -41,9 +42,37 @@ func NewSupervisor(configFile string) *Supervisor {
+	if err := s.procMgr.Update(context.TODO(), &config.Config{}); err != nil {
+		zap.L().Info("cannot terminate processes", zap.Error(err))
+	}
+	s.mu.Lock()
```
defer unlock

stuartcarnie replied:
Where is this deferred?

rogpeppe replied:
it's not! it was me in a concert with minimal connectivity perusing the code and commenting to remind myself that we must unlock the mutex here.

rogpeppe replied:
Fixed

stuartcarnie reviewed (commented):
Looks good to me, @rogpeppe – you had a comment, "defer unlock", but it isn't clear to me what that is for. I didn't see an unlock for the mutex in that function, so is it missing or is it coordinated elsewhere?

rogpeppe committed:
supervisor: add missing mutex unlock

Also, remove now-unused command line flag and don't start
a zombie reaper any more, because we don't need it because
we now wait for processes.

rogpeppe committed:
better server management

This updates the logic so that the gopm server can be shut down
and updated from within the HTTP web page.
It also tries to prevent the user from accidentally shutting
off access to their own gopm instance by starting it with
no servers or servers that can't be listened on.

It also fixes the config directory name handling so that
absolute paths work properly.

Also update the build process so that it builds a release
binary by default, so plain `go install` will do the right thing,
and change the targets so that `make` will just install the binaries
by default, and use a more specific `$GOPM_DEV` variable
for selecting a dev build (rather than just `$DEV`).

rogpeppe committed:
cmd/*: better error reports

This change prevents the entire usage message being printed when
there's a non-usage-related error, and also `gopm` now returns with a non-zero
exit code when it fails with a configuration error.

rogpeppe committed:
fix webgui code

We fix the code so it refers to the new protobuf name for the field `state`,
and add back the `description` field to the configuration, as it's shown
in the GUI so is potentially useful.

rogpeppe committed:
run CI on feature-cue branch

Also remove a couple of debugging-remnant logs.

rogpeppe committed:
move config into gopm value

We'd like a configuration to be able to contain arbitrary
fields and definitions for flexibility, so we move it down into
its own field, leaving the top level free for whatever the
user wishes.

Also fix environment variable passing to commands.

rogpeppe committed:
process: add some comments

Also remove some unused code.

rogpeppe committed:
add TODO.md

Also auto-format `config/schema.cue` and add a couple of comments.

rogpeppe committed:
config: move #RuntimeConfig out of #Config

Move the `runtime` field to the top level of the configuration
so the separation of concerns is clearer between gopm-provided
values (`runtime`) and user-specifiable values (`config`).

This avoids the name clash of having `gopm` being both a top
level field and the name of the package.

We also add a `version` field so that we can use it to disambiguate between
versions in future without worrying that some users might have defined it
themselves.

rogpeppe committed:
config: add absPath to files

The `#Paths` definition, although a bit more concise, seems
like it's likely to be a bit harder for users to understand.
Instead, add an `absPath` field to `#File` that's filled out
automatically.

Alter the filesystem test to check that it's actually possible to
refer to file paths from within programs and other files.

rogpeppe committed:
config: improve error reporting

We now report multiple-error config errors more
consistently, although it would be still better
to return them via the API.

Also remove the now-redundant group-changed members
from the `ReloadConfig` gRPC call.

Also add a `.gitattributes` file to mark the generated
Javascript files as binary so that grepping for names
of gRPC methods doesn't bring up hard-to-read results
from minified JS files.

stuartcarnie reviewed (approved):
This is a huge PR, with lots of great improvements. I haven't reviewed it in considerable depth, as I consider this largely a rewrite. You've created a TODO list and adding lots of tests which is good enough for me.

rogpeppe committed:
add dump-config command

This might make it easier to debug configuration issues
by making it possible to see the finally resolved configuration.

rogpeppe committed:
add program name prefix to shared output logs

This makes it possible to distinguish between individual
programs when logging to /dev/stdout.
